client/dashboard/dashboard-charts/index.js

```diff
@@ -18,47 +21,85 @@ import { EllipsisMenu, MenuItem, SectionHeader } from '@woocommerce/components';
  * Internal dependencies
  */
 import ChartBlock from './block';
-import { getChartFromKey, showCharts } from './config';
+import { getChartFromKey, uniqCharts } from './config';
+import withSelect from 'wc-api/with-select';
 import './style.scss';
 
 class DashboardCharts extends Component {
-	constructor() {
+	constructor( props ) {
 		super( ...arguments );
 		this.state = {
-			chartType: 'line', // @TODO: Remove this and use from props containing persisted user preferences.
-			showCharts,
+			chartType: props.userPrefChartType || 'line',
+			hiddenChartKeys: props.userPrefCharts || [],
+			query: props.query,
 		};
 
 		this.toggle = this.toggle.bind( this );
 	}
 
+	componentDidUpdate( {
+		userPrefCharts: prevUserPrefCharts,
+		userPrefChartType: prevUserPrefChartType,
+	} ) {
+		const { userPrefCharts, userPrefChartType } = this.props;
+		if ( userPrefCharts && ! isEqual( userPrefCharts, prevUserPrefCharts ) ) {
+			/* eslint-disable react/no-did-update-set-state */
+			this.setState( {
+				hiddenChartKeys: userPrefCharts,
+			} );
+			/* eslint-enable react/no-did-update-set-state */
+		}
+		if ( userPrefChartType && userPrefChartType !== prevUserPrefChartType ) {
+			/* eslint-disable react/no-did-update-set-state */
+			this.setState( {
+				chartType: userPrefChartType,
+			} );
+			/* eslint-enable react/no-did-update-set-state */
+		}
+	}
+
 	toggle( key ) {
 		return () => {
-			this.setState( state => {
-				const foundIndex = state.showCharts.findIndex( x => x.key === key );
-				state.showCharts[ foundIndex ].show = ! state.showCharts[ foundIndex ].show;
-				return state;
-			} );
+			this.setState(
+				prevState => {
+					const hidden = prevState.hiddenChartKeys.includes( key );
+					const hiddenChartKeys = prevState.hiddenChartKeys.filter( chartKey => chartKey !== key );
+
+					if ( ! hidden ) {
+						hiddenChartKeys.push( key );
+					}
+					return { hiddenChartKeys };
+				},
```

Comment:
There is a [lodash function](https://lodash.com/docs/4.17.11#xor) that implements exactly this.
```ES6
return { hiddenChartKeys: xor( prevState.hiddenChartKeys, [ key ] ) };
```

Reply:
This is awesome @Aljullu!  Thanks for the tip; had no idea.